scripts/get-flattened-publication.sh

```diff
@@ -0,0 +1,86 @@
+#!/bin/bash
+set -ueo pipefail
+
+USAGE="$0 HOST STORAGE_KEY
+Returns flat JSON corresponding with chain storage data returned by an RPC node.
+
+  HOST is a URL prefix against which an abci_query HTTP request may be issued,
+       e.g. \"https://xnet.rpc.agoric.net/\".
+  STORAGE_KEY is a chain storage key name, e.g. \"published.amm.pool0.metrics\".
+";
+
+URL_PREFIX="${1:-}";
+STORAGE_KEY="${2:-}";
+
+if [ ":${1:-}" = '--help' -o ":$URL_PREFIX" = : -o ":$STORAGE_KEY" = : -o $# -gt 2 ]; then
+  printf '%s' "$USAGE"
+  exit 64
+fi
+
+SCRIPTS_DIR="$(dirname "$(realpath "$0")")"
+FLATTENER="$SCRIPTS_DIR/flatten_json_sequence.awk"
+
+# Make the abci_query request and extract data from a response.
+# cf. https://docs.tendermint.com/master/rpc/#/ABCI/abci_query
+#
+#   {
+#     "jsonrpc": "2.0",
+#     "id": <integer>,
+#     "result": {
+#       "response": {
+#         "value": "<base64-encoded JSON text>",
+#         "height": "<decimal digits>",
+#         ...
+#       }
+#     }
+#   }
+vars="$(
+  # Avoid the GET interface in case interpretation of `path` as JSON is ever fixed.
+  # https://github.com/tendermint/tendermint/issues/9164
+  # curl -sS "${URL_PREFIX%/}/abci_query?path=%22/custom/vstorage/data/$STORAGE_KEY%22" | \
+  curl -sS "$URL_PREFIX" --request POST --header 'Content-Type: application/json' --data "$(
+    printf '{
+      "jsonrpc": "2.0",
+      "id": -1,
+      "method": "abci_query",
+      "params": { "path": "/custom/vstorage/data/%s" }
+    }' "$STORAGE_KEY"
+  )" | \
+  # Flatten into compact JSON.
+  awk -f "$FLATTENER" | \
+  # Remove the enclosing braces.
+  awk '{ sub(/^[{]/, ""); sub(/[}]$/, ""); print; }' | \
+  # Extract block height and value into shell variable assignment format (`name=value`).
+  awk -v RS=, -F: '
+    BEGIN {
+      JSON_TO_VAR["result-response-height"] = "block_height";
+      JSON_TO_VAR["result-response-value"] = "value_base64";
+    }
+    {
+      # Ignore the enclosing quotes when checking each member.
+      name = JSON_TO_VAR[substr($1, 2, length($1) - 2)];
+      if (name) {
+        # Constrain the alphabet for values.
+        if ($2 ~ /^[[:alnum:]=."_ \-]*$/) {
+          printf "%s=%s\n", name, $2;
+        }
+      }
+    }
+  '
+)"
+eval "$vars"
+
+# Decode the value into flat JSON and insert a member for the block height.
+printf '%s' "${value_base64:-}" | \
+  base64 -d | \
+  awk -f "$FLATTENER" -v unwrap=value -v capdata=true | \
```

Comment:
For the record, if `jq` is acceptable then this most complex use of `awk` (unwrap, expand capdata, and then flatten) can be replaced with the following:
```suggestion
  jq '[
    # Decode JSON in `value`, capture `slots`, then decode JSON in `body`.
    .value | fromjson | .slots as $slots | .body | fromjson |
    walk(
      if type=="object" and .["@qclass"]=="bigint" then
        # Replace bigint capdata with the sequence of digits.
        .digits
      elif type=="object" and .["@qclass"]=="slot" then
        # Replace slot reference capdata with {
        #   id: <value from global `slots`>,
        #   allegedName: <extracted from local `iface`>,
        # }.
        {
          id: $slots[.index],
          allegedName: .iface | sub("^Alleged: (?<name>.*) brand$"; "\(.name)"; "m")
        }
      else
        .
      end
    ) |
    paths(scalars) as $path |
    # Join deep object member names with "-".
    { key: $path | join("-"), value: getpath($path) } 
  ] | from_entries'
```

Reply:
I'd approve that ;-)

> if jq is acceptable

I believe it is but I defer to @arirubinstein who I think represents the customer of this script